Analysis/src/GenParticle.cpp

```diff
@@ -0,0 +1,228 @@
+        const GenParticle* mother_ = daughter->mothers.at(i);
+        if(mother_ == mother || areParented(mother_, mother))
+            return true;
+    }
+    return false;
+}
+
+const std::string& GenEvent::GetParticleName(int pdgId)
+{
+    auto iter = particle_names.find(pdgId);
+    if(iter == particle_names.end()) throw exception("Name not found for particle with pdgId = %1%") % pdgId;
+    return iter->second;
+}
+
+void GenEvent::InitializeParticleDataTable(const std::string& fileName)
+{
+    particle_names.clear();
+    particle_types.clear();
+    particle_charge.clear();
+    std::ifstream f(fileName.c_str());
+    if(!f.is_open() )
+        throw analysis::exception("Unable to read the configuration file ");
+
+    while(f.good()) {
+        std::string line;
+        std::getline(f, line);
+        if(line.empty() || line[0] == '#')
+            continue;
+        auto pdg_name_type_charge = SplitValueList(line, false, ",");
+        if(pdg_name_type_charge.size() < 2)
+            throw exception("Invalid particle definition (1): '%1%'") % line;
+
+        const int pdgId = Parse<int>(pdg_name_type_charge.at(0));
+        const std::string& name = pdg_name_type_charge.at(1);
+        if(particle_names.count(pdgId))
+            throw exception("Duplicated definition of particle with pdgId = %1%") % pdgId;
+        particle_names[pdgId] = name;
+
+        if(pdg_name_type_charge.size() >= 3)
+            particle_types[pdgId] = Parse<particles::ParticleType>(pdg_name_type_charge.at(2));
+
+
+        if(pdg_name_type_charge.size() >= 4){
+            const int& charge = Parse<int>(pdg_name_type_charge.at(3));
+            if(particle_charge.count(pdgId))
+                throw exception("Duplicated definition of particle with pdgId = %1%") % pdgId;
+            particle_charge[pdgId] = charge;
+
+        }
+    }
+}
+
+void GenEvent::PrintChain(const GenParticle* particle, const std::string& pre) const
+{
+    const int pdgParticle = particle->pdg;
+    const auto particleName = GetParticleName(pdgParticle);
+    const int particleStatus = particle->status;
+    const LorentzVectorM genParticle_momentum = particle->momentum;
+    const auto flag = particle->genStatusFlags.GenStatusFlags::getFlags();
+    auto mothers_index = particle->mothers.size() > 0 ?  particle->mothers.at(0)->index : 0;
+    std::cout << particleName                              << " <" << pdgParticle
+              << "> pt=" << genParticle_momentum.Pt()      << " eta=" << genParticle_momentum.Eta()
+              << " phi=" << genParticle_momentum.Phi()     << " E=" << genParticle_momentum.E()
+              << " m=" << genParticle_momentum.M()         << " index=" << particle->index;
+    if(particle->mothers.size() > 0){
+        for(size_t index_mother = 0; index_mother < particle->mothers.size(); ++index_mother)
+            std::cout << " mother_index=" << CollectionToString(particle->mothers, ",");
```

Comment:
I remember that you convinced me that it will not work, because mothers is vector of genParticles. So we converged to keep it as you wrote before, or I'm mistaken?